cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,917 @@
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <io/utilities/column_type_histogram.hpp>
+#include <io/utilities/parsing_utils.cuh>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+namespace {
+
+// debug accessibility
+
+// change to "\n" and 1 to make output more readable
+#define DEBUG_NEWLINE
+constexpr int DEBUG_NEWLINE_LEN = 0;
+
+// temporary? spark doesn't strictly follow the JSONPath spec.
+// I think this probably could be a configurable enum to control
+// the kind of output you get and what features are supported.
+//
+// Current known differences:
+// - When returning a string value as a single element, Spark strips the quotes.
+//   standard:   "whee"
+//   spark:      whee
+//
+// - Spark only supports the wildcard operator when in a subscript, eg  [*]
+//   It does not handle .*
+//
+// Other, non-spark known differences:
+//
+// - In jsonpath_ng, name subscripts can use double quotes instead of the standard
+//   single quotes in the query string.
+//   standard:      $.thing['subscript']
+//   jsonpath_ng:   $.thing["subscript"]
+//
+//  Currently, this code only allows single-quotes but that could be expanded if necessary.
+//
+#define SPARK_BEHAVIORS
+
+using namespace cudf;
+
+/**
+ * @brief Result of calling a parse function.
+ *
+ * The primary use of this is to distinguish between "success" and
+ * "success but no data" return cases.  For example, if you are reading the
+ * values of an array you might call a parse function in a while loop. You
+ * would want to continue doing this until you either encounter an error (parse_result::ERROR)
+ * or you get nothing back (parse_result::EMPTY)
+ */
+enum class parse_result {
+  ERROR,    // failure
+  SUCCESS,  // success
+  EMPTY,    // success, but no data
+};
+
+struct json_string {
```

Comment:
What is this? Does it represent a row element in a strings column? Is a component of the json_path? Is it the result of finding the string which the json_path addresses? Are there more than one per row? Can it really be larger than 2GB?

Reply:
It's really just a span (though I didn't think it quite warranted being made a derived class from device_span, etc) representing a string.  Use case is general purpose.  It's just a convenient carrier for passing strings around.  99% of the time, it points into some substring of the input json.   I'll document.

Reply:
Just wondering why not use `string_view`?

Reply:
Good question.  Looks like I probably could.  I'll have a go at it.

Reply:
Oh, actually, there is one issue.   Debugging this whole thing is really hard on the gpu, so I have some test code that can invoke it on the CPU.  As such, I need the class to work on host.  But for string_view, the == operator is device only.

Reply:
Sorry, but this does not sound like a good reason for this class to exist. It appears you have to change code to enable some debug support:  https://github.com/rapidsai/cudf/pull/7286/files#diff-7ddba6565b447e9b6fec6d9635c1f9f34ee8011943746eceb1674894b71c3eeeR45-R46
```
// change to "\n" and 1 to make output more readable
#define DEBUG_NEWLINE
constexpr int DEBUG_NEWLINE_LEN = 0;
```
So it seems reasonable to just add `__host__` temporarily to the `string_view` `==` operator when you want to debug this as well.
Also, I feel uneasy about adding inert host code to our compile process here.

Reply:
> So it seems reasonable to just add `__host__` temporarily to the `string_view` `==` operator when you want to debug this as well.

Doing so requires recompiling the world though.  I'll make the change, but I feel it's reasonable to have something very simple like this here for a complex piece of code.